### ISSUE TITLE
Extend uploader to support concatenation of files across lanes (e.g., HiSeq data)

### ISSUE THREAD

**Issue by boydgreenfield** [onecodex/onecodex#46]
We should add support to concatenate files in the form `*_R1_L001.fastq`, `*_R1_L002.fastq`, etc., concatenating across the `L` series and then interleaving the `R1` and `R2` reads as appropriate. 

This could _probably_ be done with a more sophisticated `_wrap_files` function (https://github.com/onecodex/onecodex/blob/master/onecodex/lib/upload.py#L42-L58) and in the `upload` routine (https://github.com/onecodex/onecodex/blob/master/onecodex/cli.py#L173).

**Comment by polyatail** [comment]
I propose implementing this by allowing users to specify multiple instances of --forward and --reverse specifying paths to multiple pairs of reads. Reads may have been generated by e.g., multiple runs on a single MiSeq instrument or multiple lanes of a HiSeq run.

We assume the user specifies reads that are paired with each other in order. Both of these are equivalent:

--forward Sample1_L001_R1.fq --forward Sample1_L002_R1.fq \
--reverse Sample1_L001_R2.fq --reverse Sample1_L002_R2.fq

--forward Sample1_L001_R1.fq --reverse Sample1_L001_R2.fq \
--forward Sample1_L002_R1.fq  --reverse Sample1_L002_R2.fq

But this command line is not:

--forward Sample1_L001_R1.fq --forward Sample1_L002_R1.fq \
--reverse Sample1_L002_R2.fq --reverse Sample1_L001_R2.fq

The app will try to verify that reads meant to be paired have matching file names, and will then call a new FASTX iterator that streams validated pairs of reads out of all sets of paired-end reads given on the command line.

**Comment by clausmith** [comment]
I think the first pass of this can actually just look for paths in the `files` argument that match (with different L00* components) and offer to concatenate across lanes in the same flow as the offer to interleave. Does that make sense? Happy to discuss. @boydgreenfield thoughts?

**Comment by polyatail** [comment]
Does it make sense to combine this with #105 #108 #109 #110?

**Comment by boydgreenfield** [comment]
Moving to backlog for discussion a bit earlier as we have a user who has received data across lanes from their sequencing provider.

**Comment by rmzelle** [comment]
We also sometimes get data from third parties across \*runs\* (e.g. if the first run didn't generate the required amount of read data), which is of course trickier to match up.

E.g. (HiSeq 2500, "123" was our provided sample ID):

> 123_S2_L001_R1_001.fastq.gz
> 123_S2_L001_R2_001.fastq.gz
> 123_S109_L003_R1_001.fastq.gz
> 123_S109_L003_R2_001.fastq.gz

Another example (NextSeq 500, "81" was our provided sample ID, different provider):

> 81_S68_R1_001.fastq.gz
> 81_S68_R2_001.fastq.gz
> 81_S93_R1_001.fastq.gz
> 81_S93_R2_001.fastq.gz

For now, it's my understanding that we can simply use `cat` to combine the fastq.gz files from different lanes/runs into a combined R1 file and a combined R2 file, before uploading those two to One Codex and letting One Codex interleave them.

**Comment by boydgreenfield** [comment]
@rmzelle That's correct – as long as you `cat` the files in the same order and include `_R1` and `_R2` in the secondary filenames, it _should_ work OK. I would err, however, towards keeping the libraries as separate samples in One Codex if they're ever from entirely different _instruments_.
